core/integration/git_test.go

```diff
@@ -768,6 +770,144 @@ func (GitSuite) TestWithAuth(ctx context.Context, t *testctx.T) {
 	})
 }
 
+func (GitSuite) TestSubmoduleAuth(ctx context.Context, t *testctx.T) {
+	c := connect(ctx, t)
+	t.Cleanup(func() { _ = c.Close() })
+
+	// Unique token per run
+	authToken := c.SetSecret("submodule-test-token", "test-token-"+uuid.NewString())
+
+	// Build content declaratively (keeps your favorite part)
+	submoduleContent := c.Directory().WithNewFile("submodule.txt", "This is the submodule content")
+	parentContent := c.Directory().WithNewFile("parent.txt", "This is the parent content")
+
+	// Test setup: bare parent + submodule repos in /srv for git-http-backend.
+	//  1. Submodule: temp worktree → commit → push to /srv/submodule.git.
+	//  2. Parent: temp worktree → commit → add submodule via absolute file:// so
+	//     `git submodule add` can fetch → rewrite .gitmodules to ../submodule.git
+	//     with `git config` → commit.
+	// Result: parent history contains a relative submodule URL; over HTTP Git
+	// resolves it relative to the parent, exercising auth-header propagation.
+
+	gitReposCtr := c.Container().
+		From(alpineImage).
+		WithExec([]string{"apk", "add", "git"}).
+		WithExec([]string{"git", "config", "--global", "user.email", "test@dagger.io"}).
+		WithExec([]string{"git", "config", "--global", "user.name", "Test User"}).
+		WithExec([]string{"git", "config", "--global", "init.defaultBranch", "main"}).
+		WithExec([]string{"sh", "-lc", "mkdir -p /srv && git init --bare /srv/submodule.git && git init --bare /srv/parent.git"}).
+		// author submodule from Dagger content, commit, push to bare
+		WithDirectory("/tmp/sub", submoduleContent).
+		WithExec([]string{"sh", "-lc", `
+    set -eux
+    cd /tmp/sub
+    git init
+    git add .
+    git commit -m 'Initial submodule commit'
+    git remote add origin file:///srv/submodule.git
+    git push -u origin main
+  `}).
+		// author parent from Dagger content, add submodule (absolute), flip to relative, commit, push
+		WithDirectory("/tmp/parent", parentContent).
+		WithExec([]string{"sh", "-lc", `
+    set -eux
+    cd /tmp/parent
+    git init
+    git add .
+    git commit -m 'Initial parent commit'
+    git -c protocol.file.allow=always submodule add file:///srv/submodule.git sub
+    git commit -m 'Add submodule (absolute url to fetch)'
+    git config -f .gitmodules submodule.sub.url ../submodule.git
+    git add .gitmodules
+    git commit -m 'Make submodule URL relative'
+    git remote add origin file:///srv/parent.git
+    git push -u origin main
+  `})
+
+	gitReposDir := gitReposCtr.Directory("/srv") // serve this via git-http-backend
+
+	// --- lighttpd smart-http with basic auth ---
+	gitService := c.Container().
```

Comment:
Question! Why aren't we using `gitServiceHTTP`?

If there's a good reason for not, then these should be split out to helpers.